src/service/url-expander/parser.js

```diff
@@ -0,0 +1,254 @@
+/**
+ * Copyright 2018 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {rethrowAsync} from '../../log';
+import {encodeValue} from '../url-replacements-impl';
+
+/** Rudamentary parser to handle nested Url replacement. */
+export class Parser {
+
+  /**
+   * Link this instance of parser to the calling UrlReplacment
+   * @param {VariableSource} variableSource the keywords to replace
+   */
+  constructor(variableSource) {
+    this.variableSource_ = variableSource;
+  }
+
+  /**
+   * Takes array of all matching variable keyword locations and returns only those that we
+   * want to evaluate. In the case of overlapping keywords it choose the longer. It also
+   * excludes protected keywords.
+   * @param {!Array<Object<string, string|number>>} matches array of objects representing
+   *  matching keywords
+   * @param {string} url The url to be expanded.
+   * @return {!Object<string, string|number>}
+   */
+  eliminateOverlaps_(matches, url, opt_whiteList) {
+    if (!matches.length) { return null; }
+    // longest keywords have priority over shorter
+    // how should we prioritize if same length??
+    matches.sort((a, b) => b.length - a.length);
+    const storage = new Array(url.length).fill(false);
```

Comment:
storage is too generic term. let's call it `keywordMarker` or so.

Reply:
changed to keywordMarker